test/test_dh.c

```diff
@@ -340,7 +340,44 @@ int test_dh_pgen_pkey(ENGINE *e, void *data)
     return err;
 }
 
-int test_dh_pkey(ENGINE *e, void *data)
+#if !defined(WE_SINGLE_THREADED) && defined(_WIN32)
+
+typedef struct {
+    ENGINE* e;
+    EVP_PKEY* params;
+} DH_KEYGEN_THREAD_VARS;
+
+/* Windows thread entry function which will test private key read access. */
+static DWORD WINAPI DhKeyGenThreadFunc(LPVOID arg)
+{
+    DH_KEYGEN_THREAD_VARS* vars = (DH_KEYGEN_THREAD_VARS*)arg;
+
+    return test_dh_pgen_pkey(vars->e, vars->params);
+}
+
+/* Regression test for problem in multi-threaded Windows environment where only
+   initial thread has private key read access while additionally created
+   threads do not */
+int test_dh_key_gen_multithreaded(ENGINE* e, EVP_PKEY* params)
+{
+    DH_KEYGEN_THREAD_VARS vars = {.e = e, .params = params};
+    HANDLE thread;
+    DWORD threadErr = 0;
+    int err = 1;
+
+    thread = CreateThread(NULL, 0, DhKeyGenThreadFunc, &vars, 0, NULL);
+    if (thread) {
+        WaitForSingleObject(thread, INFINITE);
```

Comment:
The infinite wait here could tie up the test application indefinitely, delaying or stalling any automated testing, depending on whether the automated test framework can unilaterally kill a stalled test (delay would be test framework's per-test timeout).  

Please let me know if this is a concern for our test code and I'll re-work the function to have an internal timeout value that will cause this test function to fail, allowing subsequent testing to proceed (and report failure).  Down-side to adding an internal test timeout value is that it depends on the testing environment; if set too short, can produce false positive failures...

Reply:
Yeah, I think a timeout makes sense. Agree that this test is going to be running primarily on targets with a bit of horsepower. Go ahead and add the timeout.